File: 2015/puzzle2.py
def read_puzzle():
    "read puzzle2.dat as a list of lines of integers separated by 'x'"
    with open('puzzle2.dat', 'r') as f:
        for line in f:
            yield [int(x) for x in line.strip().split('x')]

paper_size = 0

for x, y, z in read_puzzle():
    paper_size += 2 * x * y + 2 * x * z + 2 * y * z + min(x * y, x * z, y * z)
print('Part 1:', paper_size)

ribbon_size = 0

for x, y, z in read_puzzle():
    ribbon_size += min(2*x+2*y, 2*x+2*z, 2*y+2*z) + x * y * z
print('Part 2:', ribbon_size)

File: 2020/puzzle2.hs
import System.IO
import Data.List.Split (splitOn)

main :: IO ()
main = do
  lines <- readLinesFromFile "puzzle2.dat"
  let parsedLines = map parseLine (lines)
  print (length (filter isValid parsedLines))
  print (length (filter isValidPart2 parsedLines))

readLinesFromFile :: FilePath -> IO [String]
readLinesFromFile filePath = do
  contents <- readFile filePath
  return (lines contents)

parseLine :: String -> (Int, Int, Char, String)
parseLine line = (read a :: Int, read b :: Int, head c, d)
    where
        [ab, c, d] = words line
        [a, b] = splitOn "-" ab

isValid :: (Int, Int, Char, String) -> Bool
isValid (min, max, c, str) = do
  let count = length $ filter (== c) str in count >= min && count <= max

isValidPart2 :: (Int, Int, Char, String) -> Bool
isValidPart2 (pos1, pos2, c, str) = do
  let c1 = str !! (pos1 - 1)
  let c2 = str !! (pos2 - 1)
  (c1 == c && c2 /= c) || (c1 /= c && c2 == c)
